# This is Part - 01 of my final project for OCNG 669.

### This Jupyter Notebook creates an application that takes a word as an input and returns its definition (English Thesaurus).
### If the input word is not found in the database, this application will return a series of words that closely match the input.

# The function in the next cell (similarity_score) calculates "similarity score" for two words.

### The main application makes use of this function to compare the "similarity scores" of all the words in the "close matches" list.

In [1]:
#Importing the necessary libraries and the dataset for this project.

import numpy as np

#The next sequence of code imports the raw json file with a dictionary of words and definitions.
import json
data = json.load(open("C:\\Users\\rajak\\Documents\\Spring_2021\\OCNG_669\\Project\\Thesaurus Project\\data.json"))
print(type(data),len(data))

<class 'dict'> 49537


# 1. the dataset must have a minimum of 1000 data points
### This dataset has 49,537 data points; So, it satisfies the first requirement of the project.

# 2. At least one user-defined function, that takes at least one argument, and you must use the function.

### The "similarity_score" function in the next cell takes two strings as arguments and returns a similarity score based on how many letters and positions match for those two input strings.

### Since the main application makes use of this function to compare the input word to "close matches", this satisfies the second requirement of the project.

# 3. Must import one module and use a function from it

### The application already imported the json module and used the json.load method/function to read in the thesaurus dataset

### The application also imported the numpy library and uses the np.array method to create 1-D arrays of words.

In [2]:
#The next sequence creates a function that claculates the similarity score for two given strings.
#The words must be entered as strings, with " ", otherwise the function will return an error.
#However, that will not be a problem for this particular application, since we are only using this function
#inside the last for loop in the next code sequence.


def similarity_score(w1,w2):
    """ Takes two strings as arguments (include " " when writing input strings), and returns a similarity score
    based on how many letters and positions match for those two input strings."""
    lis1 = []
    lis2 = []
    for l in w1:
        lis1.append(l)
    for m in w2:
        lis2.append(m)
    i = 0
    if len(w1) < len(w2):
        for n in range(len(w1)):
            if lis1[n].lower() == lis2[n].lower():
                i += 1
        return i
    else:
        for n in range(len(w2)):
            if lis1[n].lower() == lis2[n].lower():
                i += 1
        return i


In [3]:
#This next sequence takes a word as an argument and returns its definition as the output (among other things).
#Enter \end as the input to terminate the program and start again. Otherwise you can not exit the program without
#resettig the kernel or causing an error.

while True:
    word = input("Please enter a word:")
    dummy = word
    if word == "\end":
        print("Terminated!")
        break
    elif dummy.lower() in data.keys():
        dummy1 = dummy.lower()
        output = data[dummy1]
        print(output)
    elif dummy.isnumeric():
        print("Please enter a word, not a number")
        continue
    elif (dummy.isalpha() == False):
        print("Please enter a word containing only alphabets")
        continue
    else:
        dummy1 = []
        dummy2 = []
        for i in range(1,(len(word)+1)):
            for j in list(data.keys()):
                if j.startswith(word[:i]): #and (len(j) == len(word)):
                    dummy1.append(j)
            for k in list(data.keys()):
                if k.endswith(word[-i::1]): #and (len(k) == len(word)):
                    dummy2.append(k)
        dummy3 = dummy1 + dummy2
        lis3 = []
        for p in dummy3:
            lis3.append(similarity_score(p,dummy))
        lis3 = np.array(lis3)
        dummy4 = np.array(dummy3)
        print("Did you mean to type any of the below words?\n")
        print(dummy4[lis3 >= (len(dummy)/2+1)],"\n")

Please enter a word:PHone
['An electronic device used for calling people.', 'To contact someone using the telephone.']
Please enter a word:Fhony
Did you mean to type any of the below words?

['phony' 'phony' 'phony' 'phony'] 

Please enter a word:Sarandapity
Did you mean to type any of the below words?

['Sarangani Blaan' 'Sarangani Blaan' 'Sarangani Blaan' 'Sarangani Blaan'
 'Sarangani Blaan' 'serendipity' 'serendipitously' 'serendipity'
 'serendipity' 'serendipity'] 

Please enter a word:Inmakutate
Did you mean to type any of the below words?

['inaugurate' 'immaculate' 'invalidate' 'inaugurate' 'immaculate'
 'invalidate' 'inaugurate' 'immaculate' 'invalidate'] 

Please enter a word:\end
Terminated!
